munge/srseng/dictionary.ts

```diff
@@ -0,0 +1,245 @@
+          newSlug = saferHeadWord;
+        } else {
+          for (let i = 1; ; i++) {
+            let proposed = `${saferHeadWord}@${i}`;
+            if (!used.has(proposed)) {
+              newSlug = proposed;
+              break;
+            }
+          }
+        }
+        used.add(newSlug!);
+        e.slug = newSlug;
+      }
+    }
+  }
+
+  /**
+   * Group entries by FST lemma, elect one entry to be the lemma, and demote the
+   * rest to wordforms.
+   */
+  determineLemmas() {
+    // Save locations for replacing with Wordform objects
+    const entryIndices = new Map<DictionaryEntry, number>();
+    for (let i = 0; i < this._entries.length; i++) {
+      const e = this._entries[i];
+      if (e instanceof DictionaryEntry) {
+        entryIndices.set(e, i);
+      }
+    }
+
+    // Group by lemma and lexical tags
+    const byFstLemmaAndLexicalTags = new DefaultMap<string, DictionaryEntry[]>(
+      () => Array()
+    );
+    for (const e of this._entries) {
+      if (e instanceof DictionaryEntry) {
+        if (!e.analysis) {
+          continue;
+        }
+        const fstLemma = e.fstLemma();
+        const lexicalTags = this._extractLexicalTags(e.analysis);
+        const key = JSON.stringify({ fstLemma, lexicalTags });
+        byFstLemmaAndLexicalTags.get(key).push(e);
+      }
+    }
+
+    // replace non-lemmas with wordform referring to lemmas
+    for (const [key, entries] of byFstLemmaAndLexicalTags.entries()) {
+      const { fstLemma } = JSON.parse(key);
+      const lemmaEntry = minBy(entries, (e) =>
+        stringDistance(fstLemma, e.head!)
+      );
+      assert(lemmaEntry);
+
+      // FIXME: don’t overwrite the lemma headword; requires dictionary code to
+      // first support having headword ≠ FST lemma
+      lemmaEntry.head = fstLemma;
+
+      for (const e of entries) {
+        if (e === lemmaEntry) {
+          continue;
+        }
+        const wordform = new Wordform();
+        wordform.head = e.head;
+        wordform.analysis = e.analysis;
+        wordform.senses = e.senses;
+        wordform.formOf = lemmaEntry;
+
+        const index = entryIndices.get(e);
+        assert(index);
+        entryIndices.delete(e);
+        this._entries[index] = wordform;
+      }
+    }
+  }
+
+  /**
+   * Return the set of lexical tags, suitable for use as a lookup key.
+   */
+  private _extractLexicalTags(analysis: Analysis) {
+    const tags = [...analysis[0], ...analysis[2]];
+    const ret = [];
+    for (const t of tags) {
+      if (this.lexicalTags.has(t)) {
+        ret.push(t);
+      }
+    }
+    return [...new Set(ret)].sort();
+  }
+
+  getOrCreate(text: string) {
+    assert(text);
+
+    let existing = this._byText.get(text);
+    if (existing) {
+      return existing;
+    }
+
+    const entry = new DictionaryEntry();
+
+    if (/^\p{Combining_Mark}/u.test(text.normalize())) {
+      console.log(
+        `Warning: ${JSON.stringify(text)} begins with a combining character`
+      );
```

Comment:
Does this happen often...?

Reply:
Only twice, but it’s hard to spot because the combining dot goes on the `"` at the start of the JSON string, and my terminal doesn’t show that well at all.